### ISSUE TITLE
Datashader crashes python with "noisy" timeseries data.

### ISSUE THREAD

**Issue by bardiche98** [pyviz/datashader#570]
It appears as if datashader renders low noise time series data such as the following no problem:

![low_noise](https://user-images.githubusercontent.com/35264199/37307901-3bf957de-2634-11e8-9d3b-18afd17a1acf.PNG)

However, when attempting to zoom-in on high-noise data such as the following:

![high_noise](https://user-images.githubusercontent.com/35264199/37307944-5f4ec462-2634-11e8-931c-af4d581b5902.PNG)

python bombs out with an uncaught win32 exception:

![python_crash_image](https://user-images.githubusercontent.com/35264199/37307965-7258e272-2634-11e8-9eef-b5d536127788.PNG)

I have attached a python notebook which shows this behavior (at least on my machine):

Environment:
* Windows 7
* new data shader environment (as of 2018-03-12)
* * datashader 0.6.5
* * python 3.6
* * jupyter 1.0.0

The datashader envrironment was generated using the attached yml file from the datashader examples directory.
[](url)

**Comment by bardiche98** [comment]
environment.yml content:
--------------
name: ds
channels:
 - bokeh
 - conda-forge
 - ioam

dependencies:
 - attrs
 - beautifulsoup4
 - bokeh
 - cartopy
 - colorcet
 - conda-forge::graphviz
 - conda-forge::pytest
 - conda-forge::pytest-benchmark
 - conda-forge::python-graphviz
 - dask>=0.15.4
 - datashader
 - dill
 - distributed
 - fastparquet
 - flake8
 - geoviews
 - ioam::holoviews>=1.8.3
 - ipython
 - iris
 - jupyter
 - jupyter_dashboards
 - krb5
 - matplotlib
 - nbconvert
 - nbformat
 - networkx>=2.0
 - numba
 - numpy
 - pandas
 - param>=1.5.1
 - paramnb
 - pyproj
 - pytables
 - python-snappy
 - python=3.6
 - rasterio
 - requests
 - scikit-image
 - scipy
 - shapely
 - snappy
 - statsmodels
 - tblib
 - xarray
 - yaml
 - pip:
   - cachey
   - streamz==0.2.0
   - webargs

**Comment by bardiche98** [comment]
Here is the python notebook... just remove the .txt
----------

[datashader_tseries_high_noise_error.ipynb.txt](https://github.com/bokeh/datashader/files/1804496/datashader_tseries_high_noise_error.ipynb.txt)

**Comment by jbednar** [comment]
That's a scary problem! I was able to reproduce it from your .ipynb on my Mac laptop, so it's not a Windows-specific problem.  Unfortunately, I don't see any debugging information beyond a "Kernel Restarted" message, so it will take some investigation to get to the bottom of this.  Thanks so much for reporting it!

**Comment by bardiche98** [comment]
NP!  Whew!  I thought I was going crazy for a moment. I wonder if it has to do with the percentage of the area of the plot which is being used for binning data... but that doesn't make sense since the NY taxi dataset is pretty dense... hmmm.  Anyway, thanks for the quick feedback! :-) 

 I'll be ckecking in for future updates.  Thanks again!

**Comment by jbednar** [comment]
My personal guess is that it's a memory bounds issue, and that we'll debug it by turning off Numba's jit compilation so that we get array bounds error notifications, but I won't have time to investigate it very soon. :-(  You could help a lot by narrowing it down to a non-interactive test case that reproduces it directly without user interaction, so that we can debug it outside of the Jupyter environment (to have better access to debugging tools) and so that we can reliably tell whether it has been fixed.

**Comment by bardiche98** [comment]
The frustrating thing is that it plots no problem the first time through... only bombs when you try to zoom.  Sometimes bombs on the first zoom click, othertimes on the fourth zoom click... the randomness is frustrating... But the memory bandwidth makes sense as well as I have tended to have more trouble when the number of datapoints was large...  However, a colleague noticed that it bombed with 1M points at the low noise setting.

**Comment by jbednar** [comment]
I don't think it's the bandwidth or memory size per se, just the actual values being drawn for a particular x_range, y_range value associated with that zoom revealing some problem with accessing memory outside whatever bounds there are.  That's why it would be much easier to debug if you could reduce it to a specific datashader call with one fixed range.  Obviously the initial auto-ranged version won't do it, but some other specific zoomed-in ranges will.  If you could make it display the ranges whenever you zoom interactively, you might be able to catch one that reliably produces the problem.

**Comment by bardiche98** [comment]
Here is an example notebook with logging on the x and y ranges:
---

[datashader_tseries_high_noise_error2.ipynb.txt](https://github.com/bokeh/datashader/files/1804766/datashader_tseries_high_noise_error2.ipynb.txt)

**Comment by bardiche98** [comment]
example log output:
---
2018-03-12 18:21:24,324 - root - ERROR - hello!
2018-03-12 18:21:24,324 - root - DEBUG - This is a debug message
2018-03-12 18:21:24,324 - root - INFO - this is an info message
2018-03-12 18:21:24,324 - root - WARNING - tbllalfhldfhd, warning.
2018-03-12 18:21:24,344 - root - DEBUG - x range: (1456297053.0, 1456383453.0)
2018-03-12 18:21:24,344 - root - DEBUG - y range: (-477.27164539748554, 681.0011782312909)
2018-03-12 18:21:33,084 - root - DEBUG - x range: (1456311867.9532676, 1456378503.9532676)
2018-03-12 18:21:33,084 - root - DEBUG - y range: (-477.27164539748554, 681.0011782312909)
2018-03-12 18:21:33,324 - root - DEBUG - x range: (1456311867.9532676, 1456378503.9532676)
2018-03-12 18:21:33,324 - root - DEBUG - y range: (-336.8598275323191, 556.4580876913748)
2018-03-12 18:21:35,604 - root - DEBUG - x range: (1456323293.9859753, 1456374687.0009754)
2018-03-12 18:21:35,604 - root - DEBUG - y range: (-336.8598275323191, 556.4580876913748)
2018-03-12 18:21:35,864 - root - DEBUG - x range: (1456323293.9859753, 1456374687.0009754)
2018-03-12 18:21:35,864 - root - DEBUG - y range: (-228.56721300380943, 460.4042291124646)

**Comment by bardiche98** [comment]
the last xrange / yrange combination is where it crashed.  Note that your conjecture of certain x-y range combinations makes sense now... It would crash on either 1, 2, or 3 mouse wheel zoom-events.  Looking at the logs, each of the mousewheel zoom events were of slightly different magnitudes which makes sense that the crashing behavior would be unpredictable...  Thanks!

**Comment by jbednar** [comment]
Alas, that particular combo didn't crash for me:

![image](https://user-images.githubusercontent.com/1695496/37314216-f3d2947e-2620-11e8-9505-e572fc84eadb.png)

Can you verify that on your machine it does crash if you add a cell to the end with this code?

```
x_range = (1456323293.9859753, 1456374687.0009754)
y_range = (-228.56721300380943, 460.4042291124646)

create_image(x_range, y_range, 1000, 300)
```

If that does indeed reliably crash it on your machine, I'll try to get a reproducing example the same way you did, but on my own machine.

**Comment by bardiche98** [comment]
@jbednar, Thank You for the immediate response!  Sorry that I posted the last message just before I logged out and was offline until just now.

I 
```
x_range = (1456323293.9859753, 1456374687.0009754)
y_range = (-228.56721300380943, 460.4042291124646)

create_image(x_range, y_range, 1000, 300)
```

**Comment by bardiche98** [comment]
Additionally, running it at this zoom level for the low noise plot, there was no error, but the high noise plot reliably still had the error as shown in the attached notebook.

[datashader_tseries_high_noise_error3.ipynb.txt](https://github.com/bokeh/datashader/files/1806775/datashader_tseries_high_noise_error3.ipynb.txt)

**Comment by jbednar** [comment]
Ok, thanks.  That particular range doesn't do it for my machine, but that's not too surprising, as memory overwriting depends on a lot of OS-specific factors.  Alas, I have so far been unable to get it to crash when I run your version 2 of the notebook that includes the logging commands either, so I don't currently have a way to reproduce the issue reliably myself.  I did see it once a couple of days ago, but never since!

**Comment by bardiche98** [comment]
Hmmm... Have you tried turning on the wheel zoom and just randomly zooming in and out all over the place?  This is where I first noticed the problem with smaller data sets (or maybe that is me just trying to assign patterns where none exist... ).

It also sort of seemed to be weakly linked to data set size so maybe bump up the data set size until it hogs most available ram without crashing python on the initial data load.  And then try zooming?

**Comment by jbednar** [comment]
Yes, that's what I tried, which worked initially, but for some reason not today...

**Comment by bardiche98** [comment]
Additional note that it fails on both Win 7 and Win 10, with two different hardware manufacturers.

**Comment by bardiche98** [comment]
Here is the output from the debugger... unfortunately, my python binary wasn't compiled with debugger mode on... (at least that is what the debugger told me)...  Note that the value of r10 and rax on different failure instances (both to different and identical zooming levels) were different.  

![debugger](https://user-images.githubusercontent.com/35264199/37488670-4e72aa9a-288d-11e8-8e77-4ee746193732.PNG)

**Comment by bardiche98** [comment]
Note that when we tried this notebook on a MacBook Pro, we also did not get it to crash.  @jbednar, is there anything else I can do to help troubleshoot this one?  Thanks!

**Comment by kgullikson88** [comment]
I'm not sure if its the same issue, but I've noticed that the kernel often crashes when your data spans many orders of magnitude. Adding these two lines of code to the cell making the dataframe causes it to crashes as soon as I try to plot:

```python
samp = df.sample(frac=0.1)
df.loc[samp.index, 'data'] = 1e9
```

[datashader_tseries_outlier_error.ipynb.txt](https://github.com/bokeh/datashader/files/1921515/datashader_tseries_outlier_error.ipynb.txt)

**Comment by jbednar** [comment]
Thanks for that observation; it might help!

**Comment by alexlee93** [comment]
I'm having the same problem, but my case it is due to the issue identified by @kgullikson88 above. I'm using Datashader with Holoviews and running the app through Bokeh server in Chrome
Environment:
- Mac OS 10.13.14
- Python version: 3.6.5
- Bokeh version: 0.12.16
- Datashader version: 0.6.5
- Holoviews version: 1.10.4

**Comment by kgullikson88** [comment]
@alexlee93 yes, I have the same problem on my setup with bokeh server, doing almost identical thing to you. The issue generally seems to be worse on bokeh-server than it is in the notebook for me.

**Comment by KBlackbird** [comment]
I came across this thread because I am having similar issues with bokeh+datashader+holoviews. One thing I did note is that if I change the plot type from hv.Curve to hv.Scatter, then I do not get the crashes anymore. It's not the visualization I want, but I can abuse the mouse zoom pretty well and I haven't crashed it yet.

**Comment by jbednar** [comment]
No progress yet, but we are looking at it.

**Comment by jbednar** [comment]
That's super helpful, thanks! We've been distracted by other tasks, but this will really make it simple to fix!

**Comment by ceball** [comment]
I'm not familiar with this bit of the code, and I couldn't reproduce the crashes myself, but the very helpful information above was enough to find a suspicious pixel index calculation. I modified it in a similar way to the suggestion above, although not in exactly the same way. Could someone who currently gets crashes with lines see if #644 fixes things?

**Comment by jbednar** [comment]
For what it's worth, I was able to reproduce the original issue using Datashader master and my typical environment, so at least the problem itself is still relevant:
![image](https://user-images.githubusercontent.com/1695496/45173412-c0736480-b1cd-11e8-92b0-706aaf30a71b.png)
But it took quite a bit of zooming around, and I changed the noise_constant from 100 to 1000, which might have been important.  I'll try now with the PR applied, though not finding any problem won't necessarily indicate that it's gone, because it did take several minutes of zooming to make it happen with master.

**Comment by jbednar** [comment]
Well, it's hard to interpret a negative result, but I spent 5 minutes zooming into that same notebook every which way after pulling the changes on your PR, and never had an issue.  Plus the fix to the code seems very appropriate; I can see how it would go wrong in some cases but usually not be a problem.  So I think you should make that PR mergeable and then merge it.

**Comment by bardiche98** [comment]
@jbednar & @philippjfr, sorry for the long radio silence on this one...  

I just created a new environment with bokeh 0.13.0 and datashader 0.6.8 (which has the bug fix) with and then tried the jupyter notebook provided above `datashader_tseries_high_noise_error3.ipynb`, and found that the win32 uncaught exception still exists for the example provided in the notebook.

**Comment by bardiche98** [comment]
Note that patching the return statement in map_onto_pixel() to check for < the boundary instead of the point lying on the boundary per @p3trus fixes the issue for me.

```
def _build_map_onto_pixel_for_line(x_mapper, y_mapper):
    @ngjit
    def map_onto_pixel(vt, bounds, x, y):
        """Map points onto pixel grid.

        Points falling on upper bound are mapped into previous bin.

        If the line has been clipped, x and y will have been
        computed to lie on the bounds; we compare point and bounds
        in integer space to avoid fp error. In contrast, with
        auto-ranging, a point on the bounds will be the same
        floating point number as the bound, so comparison in fp
        representation of continuous space or in integer space
        doesn't change anything.
        """
        sx, tx, sy, ty = vt
        xmax, ymax = bounds[1], bounds[3]
        xx = int(x_mapper(x) * sx + tx)
        yy = int(y_mapper(y) * sy + ty)

        xxmax = int(x_mapper(xmax) * sx + tx)
        yymax = int(y_mapper(ymax) * sy + ty)

        # datashader v0.6.8 return statement
        # return (xx - 1 if xx == xxmax else xx,
        #         yy - 1 if yy == yymax else yy)

        # patched return statement which fixes this Issue #570
        return (xx if xx < xxmax else xxmax - 1, yy if yy < yymax else yymax - 1)

    return map_onto_pixel
```

Note that there are two other similar return statements in `glyphs.py` which check for the point on the boundary instead of point beyond the boundary; however, patching those return statements do not require patching to fix this issue.  While they do not require patching for this present issue, it probably doesn't hurt to also have them check if the point is on or beyond the boundary as well.

**Comment by bardiche98** [comment]
Agreed with @ceball.  

With the caveat that I have not looked at the clipping code, this proposed fix seems like it will be OK for time-series data as long as the point-to-point time delta is not very many pixels and would not produce noticeable incorrect visualizations; however, is not the right solution and could have unintended side-effects for line-scatter-plots as as discussed below:

Since it seems like the points are beyond the plotting range and we are wanting to draw a line to it, we would want to ensure that the point being clipped is not just translated vertically or horizontally to the boundary for plotting (which is what the proposed fix above does by hiding the underlying issue), it should be translated along the line connecting the two points prior to being plotted on the boundary.  In other words, need to find the intersection of the boundary and the line connecting the two points and end the line we are drawing at the intersection.   If this is not done, there will be artifacts which occur at the boundaries, and the lines which connect to points off of the image will change slope as you zoom in and out as the boundary moves.  This likely wouldn't impact time-series data too bad, but could impact line-scatter plots or line-scatter-plots without very many points in the visible area.
